File: application/backend-credit-card-enrollment/backend-c#/CreditCard/Enrollment/Command/RequestEnrollmentCommandHandler.cs
using CreditCardEnrollment.Common.Command;
using CreditCardEnrollment.Common.EventStore;
using CreditCardEnrollment.Common.SessionAuth;
using CreditCardEnrollment.Common.Util;
using CreditCardEnrollment.CreditCard.Enrollment.Event;
using CreditCardEnrollment.CreditCard.Enrollment.Projection.IsProductActive;

namespace CreditCardEnrollment.CreditCard.Enrollment.Command;

public class RequestEnrollmentCommandHandler : CommandHandler {
    private readonly SessionService _sessionService;
    private readonly IsProductActive _isProductActive;

    public RequestEnrollmentCommandHandler(
        PostgresTransactionalEventStore postgresTransactionalEventStore,
        SessionService sessionService,
        IsProductActive isProductActive) : base(postgresTransactionalEventStore) {
        _sessionService = sessionService;
        _isProductActive = isProductActive;
    }

    public override void HandleCommand(Common.Command.Command command) {
        if (command is RequestEnrollmentCommand enrollmentCommand) {
            HandleRequestEnrollment(enrollmentCommand);
        } else {
            throw new ArgumentException($"Unsupported command type: {command.GetType().Name}");
        }
    }

    private void HandleRequestEnrollment(RequestEnrollmentCommand command) {
        var userId = _sessionService.AuthenticatedUserIdFromSessionToken(command.SessionToken);

        if (!_isProductActive.IsProductActiveById(command.ProductId)) {
            throw new Exception("Product is inactive and not eligible for enrollment request.");
        }

        var eventId = IdGenerator.GenerateRandomId();
        var aggregateId = IdGenerator.GenerateRandomId();
        var enrollmentRequested = new EnrollmentRequested {
            EventId = eventId,
            AggregateId = aggregateId,
            AggregateVersion = 1,
            CorrelationId = eventId,
            CausationId = eventId,
            RecordedOn = DateTime.UtcNow,
            UserId = userId,
            ProductId = command.ProductId,
            AnnualIncomeInCents = command.AnnualIncomeInCents
        };

        _postgresTransactionalEventStore.SaveEvent(enrollmentRequested);
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/creditCard/enrollment/command/RequestEnrollmentCommandHandler.ts
import { CommandHandler } from '../../../common/command/CommandHandler';
import { PostgresTransactionalEventStore } from '../../../common/eventStore/PostgresTransactionalEventStore';
import { IdGenerator } from '../../../common/util/IdGenerator';
import { RequestEnrollmentCommand } from './RequestEnrollmentCommand';
import { EnrollmentRequested } from '../event/EnrollmentRequested';
import {SessionService} from "../../../common/sessionAuth/SessionService";
import {IsProductActive} from "../projection/isProductActive/IsProductActive";

export class RequestEnrollmentCommandHandler extends CommandHandler {
    private readonly sessionService: SessionService;
    private readonly isProductActive: IsProductActive;

    constructor(
        postgresTransactionalEventStore: PostgresTransactionalEventStore,
        sessionService: SessionService,
        isProductActive: IsProductActive
    ) {
        super(postgresTransactionalEventStore);
        this.sessionService = sessionService;
        this.isProductActive = isProductActive;
    }

    async handleCommand(command: RequestEnrollmentCommand): Promise<void> {
        const userId = await this.sessionService.authenticatedUserIdFromSessionToken(command.sessionToken);

        if (!(await this.isProductActive.isProductActiveById(command.productId))) {
            throw new Error('Product is inactive and not eligible for enrollment request.');
        }

        const eventId = IdGenerator.generateRandomId();
        const aggregateId = IdGenerator.generateRandomId();

        const enrollmentRequested = new EnrollmentRequested(
            eventId,
            aggregateId,
            1,
            eventId,
            eventId,
            new Date(),
            userId,
            command.productId,
            command.annualIncomeInCents
        );

        await this.postgresTransactionalEventStore.saveEvent(enrollmentRequested);
    }
}
